### PR TITLE
Call context

### PR DESCRIPTION
**Pull request by software-dov**

### CODE REVIEW

File: gax/call_context.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+#include <map>
+#include <string>
+#include <utility>
+
+using GrpcContextPolicyFunc = std::function<void(grpc::ClientContext*)>;
+
+namespace google {
+namespace gax {
+
+struct MethodInfo {
```
Comment (michaelbausor) [inline gax/call_context.h]: Can we add a comment to this struct?

File: gax/call_context.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+ *                               Foo* response) {
+ *   grpc::ClientContext c;
+ *   context.PrepareGrpcContext(&c);
+ *   return grpc_stub_->GetFoo(&c, request, response);
+ * }
+ */
+class CallContext {
+ public:
+  CallContext(MethodInfo method_info)
+      : deadline_(std::chrono::time_point<std::chrono::system_clock>::max()),
```
Comment (michaelbausor) [inline gax/call_context.h]: Why not `std::chrono::system_clock::time_point`, as used elsewhere in this file?

File: gax/call_context.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+#include "grpcpp/client_context.h"
+#include "internal/gtest_prod.h"
+
+#include <chrono>
+#include <functional>
+#include <map>
+#include <string>
+#include <utility>
+
+using GrpcContextPolicyFunc = std::function<void(grpc::ClientContext*)>;
```
Comment (michaelbausor) [inline gax/call_context.h]: Can we add a comment to document this signature?

File: gax/call_context.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+   * This function could tweak advanced knobs or provide other custom behavior.
+   */
+  void AddGrpcContextPolicy(GrpcContextPolicyFunc f);
+
+  /**
+   * Initializes a grpc::ClientContext, adding stored metadata, setting a
+   * deadline, and applying registered customization functions.
+   *
+   * @par Pre-conditions
+   * The 'context' parameter is pristine and has not been used for an rpc.
```
Comment (michaelbausor) [inline gax/call_context.h]: Nit: it doesn't necessarily need to be 'pristine' (maybe there will be a case when a CallContext is used on an existing grpc::ClientContext that had already been partially configured, although I can't think of any such case). But it does need to not have been used for an rpc.

File: gax/call_context.h [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  /**
+   * @name Accessor for method info.
+   */
+  MethodInfo Info() const { return method_info_; }
+
+ private:
+  FRIEND_TEST(CallContext, Basic);
+  std::chrono::system_clock::time_point deadline_;
+  std::vector<GrpcContextPolicyFunc> context_policies_;
+  std::map<std::string const, std::string const> metadata_;
```
Comment (michaelbausor) [inline gax/call_context.h]: This needs to be a multimap - it is valid to register the same key multiple times.

File: gax/call_context.h (hunk 4 above)
Comment (software-dov) [reply gax/call_context.h]: Good point, a pristine ClientContext isn't a hard prerequisite, but it makes it much easier to reason about the metadata. I'll update the comment.

File: gax/call_context.h (hunk 5 above)
Comment (software-dov) [reply gax/call_context.h]: Done

File: gax/call_context.h (hunk 3 above)
Comment (software-dov) [reply gax/call_context.h]: Done

File: gax/call_context.h (hunk 1 above)
Comment (software-dov) [reply gax/call_context.h]: Done

File: gax/call_context.h (hunk 2 above)
Comment (software-dov) [reply gax/call_context.h]: I had been getting template-based compilation errors when trying to use `std::chrono::system_clock`. These seem to have gone away, so changing the invocation of `max()` to be consistent with the rest of the file.

File: WORKSPACE [hunk 6]
```diff
@@ -16,13 +16,19 @@ workspace(name = "com_google_gapic_generator_cpp")
...
 
+com_google_gapic_generator_cpp_gax_repositories()
 com_google_gapic_generator_cpp_repositories()
 
 bind(
     name = "zlib",
     actual = "@net_zlib//:zlib",
 )
 
+# NOTE: this is a hack to allow grpc to build
```
Comment (michaelbausor) [inline WORKSPACE]: Does this go away once https://github.com/googleapis/gapic-generator-cpp/pull/31 is merged?

File: gax/call_context_unittest.cc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+              "CallContext must be move constructable");
+static_assert(std::is_copy_constructible<CallContext>::value,
+              "CallContext must be copy constructable");
+
+TEST(CallContext, Basic) {
+  gax::MethodInfo mi{"TestMethod", MethodInfo::RpcType::CLIENT_STREAMING,
+                     MethodInfo::Idempotency::IDEMPOTENT};
+  gax::CallContext ctx(mi);
+  EXPECT_EQ(ctx.Deadline(),
+            std::chrono::time_point<std::chrono::system_clock>::max());
```
Comment (michaelbausor) [inline gax/call_context_unittest.cc]: Nit: std::chrono::system_clock::time_point

File: WORKSPACE (hunk 6 above)
Comment (software-dov) [reply WORKSPACE]: Wait a minute, I'll go check.

File: gax/call_context_unittest.cc (hunk 7 above)
Comment (software-dov) [reply gax/call_context_unittest.cc]: Done

File: WORKSPACE (hunk 6 above)
Comment (software-dov) [reply WORKSPACE]: Yes. It's a known limitation of bazel.

Comment (michaelbausor) [reply WORKSPACE]: Wait, yes it goes away? Or yes it is still needed?

Comment (software-dov) [reply WORKSPACE]: Aaah! Bad phrasing/hasty reading on my part. Yes, it's still needed; no, it does not go away.

File: gax/call_context.h [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  /**
+   * @name Accessor for method info.
+   */
+  MethodInfo Info() const;
+
+ private:
+  FRIEND_TEST(CallContext, Basic);
+  std::chrono::system_clock::time_point deadline_;
+  std::vector<GrpcContextPolicyFunc> context_policies_;
+  std::multimap<std::string const, std::string const> metadata_;
```
Comment (devbww) [inline gax/call_context.h]: Note that const qualifying the key of a map is generally discouraged.  The key is already immutable, so it doesn't buy you anything.

Comment (software-dov) [reply gax/call_context.h]: Good to know. I wasn't sure about the guarantees: mutating the key seems like a really, really, _really_ bad idea with no good justification, but this is C++.

**Pull request merged**